docs/bls-keys/dd.md

```diff
@@ -0,0 +1,103 @@
+# Validator BLS Keys Design Document
+
+- Owner: @spalladino
+- Approvers: @LHerskind, @iAmMichaelConnor
+
+## Summary
+
+We have identified potential gas savings from aggregating signatures, both for L2 block proposals (by aggregating block attestations) and for voting (by aggregating votes). Regardless of whether we implement these gas savings now, we want to freeze governance code soon, so we propose future-proofing it by already adding support for validators to register BLS keys.
+
+## Signing scheme and curve
+
+The requirements for the signing scheme and curve are:
+
+- An aggregated signature should be small (no more than 512 bits ideally) for a 48-validator committee.
+- The cost of computing an aggregate key and verifying an aggregate signature should not be too expensive in the EVM (no more than 300k gas ideally). Note that, since the committee changes once per epoch, we need to compute a new aggregate key every epoch.
+- Each individual public key should be small as well, since it will have to be read from storage for computing the committee commitment.
+- At the moment, we do not require verifying these signatures on a rollup circuit, but this _may_ change in the future, so it should not be prohibitively expensive.
+
+We propose that a BLS aggregate signature over the [BLS12-381](https://eth2book.info/latest/part2/building_blocks/bls12-381/) curve fits the bill. Public keys are 48 bytes long, aggregated signatures are 96 bytes long, and L1 verification should be doable in under the target gas:
+
+> We need to send as CALLDATA the 48 pubkeys for the committee members plus a bitmap of signers, so we can reconstruct both the committee hash and the aggregated pubkey, which is roughly 48 words (`1536` bytes, `61k` gas at 40 gas per slot, or `24k` at 16 gas per slot). The verification itself requires two `SLOAD`s (`4200`), a hash-to-curve (`20k` gas), 33 ECADDs (`5k`), and two pairings (`124k` gas). Total is about `214k` gas, which gets paid once per epoch.
+
+## Data types
+
+We add a `BLSPublicKey` type, backed 48 bytes data, to fit a public key on the BLS12-381 curve. Along with this type, we add a method for validating that a given `BLSPublicKey` instance is a valid point in the curve.
+
+## Interface changes
+
+The following methods or structs require an additional `BLSPublicKey` field:
+
+- `StakingQueue.enqueue`
+- `StakingQueue.DepositArgs`
+- `IStakingCore.deposit`
+- `StakingLib.deposit`
+- `RollupCore.deposit`
+- `ExtRollupLib2.deposit`
+- `GSE.deposit`
+- `GSE.AttesterConfig`
+
+All GSE functions that return a set of attester addresses now require an additional function for fetching their BLS keys:
+
+- `GSE.getAttestersFromIndicesAtTime` -> `GSE.getAttestersBLSPubKeysFromIndicesAtTime`
+- `GSE.getAttestersAtTime` -> `GSE.getAttestersBLSPubKeysAtTime`
+- `GSE.getAttesterFromIndexAtTime` -> `GSE.getAttesterBLSPubKeyFromIndexAtTime`
+
+## Security considerations
```

Comment:
Would also expect that there can be some somewhat weird quirks if we are doing simple key aggregation with an attacker registering `a` and `-a` so when both are in the committee it effectively reduces the size. Essentially it gives a "fake" safety, as while there might be funds at stake for those, anyone can use it, so it is not bringing much safety. Generally some weird quicks there, so the 2'nd approach mentioned might be a must. Need someone with better math skills than me to look.